src/dbami/helpers/apply_migrations.py

```diff
@@ -0,0 +1,213 @@
+    #     close will fire and proceed without any guarantee of success.
+    #
+    #     This setting only applies if force_close_connections_after_wait is
+    #     set to True.
+    connection_wait_poll_interval_ms: int = 2000
+    connection_wait_max_attempts: Optional[int] = None
+    force_close_connections_after_wait: bool = False
+    force_close_connections_timeout_ms: int = 10000
+
+    def __post_init__(self) -> None:
+        if self.connection_wait_poll_interval_ms < 100:
+            self.connection_wait_poll_interval_ms = 100
+
+    @property
+    def wait_for_other_connections_to_close(self) -> bool:
+        return (
+            self.connection_wait_max_attempts is None
+            or self.connection_wait_max_attempts > 0
+        )
+
+
+class MigrationHelper:
+    def __init__(
+        self,
+        database: DB,
+        helper_config: Optional[MigrationHelperConfig] = None,
+        logger: logging.Logger = logger,
+        **connect_kwargs,
+    ) -> None:
+        self.database = database
+        self.logger = logger
+        self.connect_kwargs = connect_kwargs
+
+        if helper_config is None:
+            helper_config = MigrationHelperConfig()
+
+        self.config = helper_config
+
+
+    async def grant_connect_privileges(
+        self,
+        conn: asyncpg.Connection,
+    ) -> None:
+        for role_name in self.config.revoke_connect_on_roles_names:
+            self.logger.warning("Granting connect privileges for role '%s'", role_name)
+            q, p = render(
+                """
+                    DO $_$
+                        BEGIN
+                            EXECUTE FORMAT('GRANT CONNECT on database %s TO :role',
+                            CURRENT_DATABASE());
+                        END
+                    $_$;
+                """,
+                role=V(role_name),
+            )
+            await conn.execute(q, *p)
+
+
+    async def revoke_connect_privileges(
+        self,
+        conn: asyncpg.Connection,
+    ) -> None:
+        for role_name in self.config.revoke_connect_on_roles_names:
+            self.logger.warning("Revoking connect privileges for role '%s'", role_name)
+            q, p = render(
+                """
+                    DO $_$
+                        BEGIN
+                            EXECUTE FORMAT('REVOKE CONNECT on database %s FROM :role',
+                            CURRENT_DATABASE());
+                        END
+                    $_$;
+                """,
+                role=V(role_name),
+            )
+            await conn.execute(q, *p)
+
+
+    async def active_connections_exist(self, conn: asyncpg.Connection) -> bool:
+        return bool(
+            await conn.fetchval(
+                """
+            SELECT EXISTS(
+                SELECT * FROM pg_stat_activity
+                WHERE
+                    datname = current_database()
+                    AND pid != pg_backend_pid()
```

Comment:
This is wrong, as was pointed out on a previous revision in a different comment. This likely should be matching on the role name of the application role, to see if we have any application connections. As it stands right now it will return true if any other processes are connecting with the owner role to also try to apply migrations, which is fine to allow (due to locking that we need to add in the migration apply).

Reply:
Consider also `AND usename != current_user`.

Reply:
added